crates/modelardb_storage/src/lib.rs

```diff
@@ -311,13 +321,264 @@ fn apache_parquet_writer_properties(
+        if let Some(generated_column) = generated_column {
+            if let Some(sql_expr) = generated_column.original_expr {
+                generated_columns_builder.values().append_value(sql_expr);
+            } else {
+                generated_columns_builder.values().append_null();
+            }
+        } else {
+            generated_columns_builder.values().append_null();
+        }
+    }
+
+    generated_columns_builder.append(true);
+    generated_columns_builder.finish()
+}
+
+/// Extract the table metadata from `record_batch` and return the table metadata as a tuple of
+/// `(normal_table_metadata, model_table_metadata)`. `normal_table_metadata` is a vector of tuples
+/// containing the table name and schema of the normal tables. If the schema of the record batch
+/// is invalid or the table metadata could not be extracted, return [`ModelarDbStorageError`].
+#[allow(clippy::type_complexity)]
+pub fn table_metadata_from_record_batch(
+    record_batch: &RecordBatch,
+) -> Result<(Vec<(String, Schema)>, Vec<ModelTableMetadata>)> {
+    if record_batch.schema() != TABLE_METADATA_SCHEMA.0 {
+        return Err(ModelarDbStorageError::InvalidArgument(
+            "Record batch does not contain the expected table metadata.".to_owned(),
+        ));
+    }
+
+    let mut normal_table_metadata = Vec::new();
+    let mut model_table_metadata = Vec::new();
+
+    let type_array = modelardb_types::array!(record_batch, 0, StringArray);
+    let name_array = modelardb_types::array!(record_batch, 1, StringArray);
+    let schema_array = modelardb_types::array!(record_batch, 2, BinaryArray);
+    let error_bounds_array = modelardb_types::array!(record_batch, 3, ListArray);
+    let generated_columns_array = modelardb_types::array!(record_batch, 4, ListArray);
+
+    for row_index in 0..record_batch.num_rows() {
+        let table_type = type_array.value(row_index);
+        let table_name = name_array.value(row_index).to_owned();
+        let schema = try_convert_bytes_to_schema(schema_array.value(row_index).to_vec())?;
+
+        match table_type {
+            "normal" => normal_table_metadata.push((table_name, schema)),
+            "model" => {
+                let error_bounds = array_to_error_bounds(error_bounds_array.value(row_index))?;
+
+                let generated_columns = array_to_generated_columns(
+                    generated_columns_array.value(row_index),
+                    &schema.clone().to_dfschema()?,
+                )?;
+
+                let metadata = ModelTableMetadata::try_new(
+                    table_name,
+                    Arc::new(schema),
+                    error_bounds,
+                    generated_columns,
+                )?;
+
+                model_table_metadata.push(metadata);
+            }
+            _ => {
+                return Err(ModelarDbStorageError::InvalidArgument(format!(
+                    "Table type '{table_type}' is not supported.",
```

Comment:
Should a boolean array be used for table type to make it impossible to represent anything but normal tables and model tables?

Reply:
Good idea, fixed.

Reply:
Good idea, fixed.